core/env.go

```diff
@@ -0,0 +1,128 @@
+package core
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+)
+
+// Env implementations provide the environment for maintaining bindings
+// and evaluating forms.
+type Env interface {
+	Eval(form Value) (Value, error)
+	Bind(symbol string, v Value) error
+	Resolve(symbol string) (Value, error)
+	Parent() Env
+}
+
+// New returns an implementation of env with no bindings.
+func New(parent Env) Env {
+	return &mapEnv{
+		mu:     &sync.RWMutex{},
+		scope:  map[string]Value{},
+		parent: parent,
+	}
+}
+
+type mapEnv struct {
+	mu     *sync.RWMutex
```

Comment:
Yea, makes sense. (I am always scared of the enclosing struct being passed by value and `mu` getting copied. Documentation explicitly says _"A RWMutex must not be copied after first use."_ . So in most cases i simply use the pointer value. But in this case since `mapEnv` is hidden behind interface and uses pointer receiver it should be okay.)

Reply:
>I am always scared of the enclosing struct being passed by value and mu getting copied.

For what it's worth, linters will pick this up.  I tend not to worry about this too much.

>in this case since mapEnv is hidden behind interface and uses pointer receiver it should be okay

Yes.  See also item 5 in my long comment below.  😃